### ISSUE TITLE
ICP dobbler switch does not work in BMS 4.36

### ISSUE THREAD

**Issue by gabrielwong1991** [kungfoo/falcon-bms-control#21]
Hi for some reason the ICP dobber switch (Left, Right, Up, Down) does not work in BMS 4.36.

Maybe there is a new memory address/offset for these keys in the new version?

Thanks

**Comment by kungfoo** [comment]
Thank you for reporting this issue.

I'll see if i can reproduce this tomorrow.

**Comment by kungfoo** [comment]
Tested it again on BMS 4.36: _It works fine on my machine._

I'm pretty sure you probably don't have a keybind (keyboard) for those functions in your keyfile.
Do you or did you have any DX Device bound to those functions? Are you using Alternate Launcher and some sort of pre-defined profile?

The server puts a log statement about every callback issued and a warning about the ones, where there is no keybind that is found. For now, that is the best I can do, since the server part _has to pretend to be a keyboard._

**Comment by gabrielwong1991** [comment]
Hmm... I am currently using X52 Pro - Falconeer v1.2 PROFILE.

The log file does not seemed to warn anything related to Up down left right key binding. For example the Up and Down log

```2022-05-02 20:59:29.6737|DEBUG|CallbackSender|Down SimICPDEDUP
2022-05-02 20:59:29.7257|DEBUG|IcpButtonHandler|icp-button: icp-released:icp-ded-up
2022-05-02 20:59:29.7257|DEBUG|CallbackSender|Released SimICPDEDUP
2022-05-02 20:59:30.3746|DEBUG|IcpButtonHandler|icp-button: icp-pressed:icp-ded-down
2022-05-02 20:59:30.3746|DEBUG|CallbackSender|Down SimICPDEDDOWN```

**Comment by kungfoo** [comment]
That is a bit odd. Can you make sure you run the newest server version from here: https://github.com/kungfoo/falcon-bms-control/releases/tag/server-0.5 and check the log again?

You could try and find all the bindings for that callback in your keyfile by searching for `SimICPDEDDOWN` and see if there is only DX bindings or also a keyboard binding. I'll have to take a look at that profile later.

**Comment by kungfoo** [comment]
I downloaded the profile from here [VG x52 Pro Falconeer Joystick Profile](https://veterans-gaming.com/files/file/38-x52-pro-falconeer-joystick-profile-for-bms/) and checked the keyfile. It looks as if the keybinds should be there...

Can you confirm, that you have those lines on your keyfile?

```
SimICPDEDUP 120 0 0xC8 0 0 0 1 "ICP: DCS UP - Push"
SimICPDEDDOWN 120 0 0xD0 0 0 0 1 "ICP: DCS DOWN - Push"
SimICPDEDSEQ 120 0 0xCD 0 0 0 1 "ICP: DCS SEQ (Right) - Push"
SimICPResetDED 120 0 0xCB 0 0 0 1 "ICP: DCS RTN (Left) - Push"
```

**Comment by gabrielwong1991** [comment]
Hi kungfoo,

I am using version 0.5 for the server and had a look at the profile and indeed the lines are there:

![image](https://user-images.githubusercontent.com/30657000/166937357-cd4ae000-9d56-4d31-b4d9-81d7175076b2.png)

**Comment by gabrielwong1991** [comment]
I also discovered some wierd behaviour (it works as it intended to) is that when you click on the right MFD OBS11 to 20 i will see the language bar pops up like this:

https://photos.google.com/photo/AF1QipPEiz91m7ALFKRnVRAG_GFdw5B8pLHmVdUODFNY

Though it will works as it intended to.

**Comment by kungfoo** [comment]
Oh, I've heard that problem from people using multiple input languages as well. If you can live with it, you could remove one of the languages or disable the language bar.

One more question: can you check if setting the language to `US` fixes the problem with the dobber switch?

**Comment by gabrielwong1991** [comment]
I removed UK as input language and US is now primary, the language bar thing is gone but it still does not solves the dobber switch problem...

**Comment by gabrielwong1991** [comment]
Hey kungfoo, turns out the falconeer key file's ICP up, down, left, right is colliding with Radar mini cursor!

Originally ICP up, down, left, right is the normal arrow keys in keyboard and radar mini cursor is arrow keys + shift key...

Turns out X52's internal profile cannot add shift + arrow keys to the Radar mini cursor so Falconeer's one didnt add any modifier (without shift)

So is there a way we can customise in your program to change the ICP up, down, left ,right keys with shift?

--------------------------------------------------------------------------------------------------------------------------------------

Another question is, I am using Ipad as the client app and i found that the program disconnect (freezes) consistently after, say, 10 minutes of use. I tried increasing compression and lower update rate with no use... When it tries to reconnect it just won't...

Though the ipad feels hot after 10 mins of use and I wondered if Ipad OS have some internal mechanisms that maybe reduce transfter rate when it is hot...  not sure
